src/System.IO/src/System/IO/BufferedStream.cs

```diff
@@ -0,0 +1,682 @@
+            if (_writePos >= _bufferSize - 1)
+                FlushWrite();
+
+            _buffer[_writePos++] = value;
+
+            Contract.Assert(_writePos < _bufferSize);
+        }
+
+        public override Int64 Seek(Int64 offset, SeekOrigin origin)
+        {
+            EnsureNotClosed();
+            EnsureCanSeek();
+
+            // If we have bytes in the WRITE buffer, flush them out, seek and be done.
+            if (_writePos > 0)
+            {
+                // We should be only writing the buffer and not flushing,
+                // but the previous version did flush and we stick to it for back-compat reasons.
+                FlushWrite();
+                return _stream.Seek(offset, origin);
+            }
+
+            // The buffer is either empty or we have a buffered READ.
+
+            if (_readLen - _readPos > 0 && origin == SeekOrigin.Current)
+            {
+                // If we have bytes in the READ buffer, adjust the seek offset to account for the resulting difference
+                // between this stream's position and the underlying stream's position.            
+                offset -= (_readLen - _readPos);
+            }
+
+            Int64 oldPos = Position;
+            Contract.Assert(oldPos == _stream.Position + (_readPos - _readLen));
+
+            Int64 newPos = _stream.Seek(offset, origin);
+
+            // If the seek destination is still within the data currently in the buffer, we want to keep the buffer data and continue using it.
+            // Otherwise we will throw away the buffer. This can only happen on READ, as we flushed WRITE data above.
+
+            // The offset of the new/updated seek pointer within _buffer:
+            _readPos = (Int32)(newPos - (oldPos - _readPos));
+
+            // If the offset of the updated seek pointer in the buffer is still legal, then we can keep using the buffer:
+            if (0 <= _readPos && _readPos < _readLen)
+            {
+                // Adjust the seek pointer of the underlying stream to reflect the amount of useful bytes in the read buffer:
+                _stream.Seek(_readLen - _readPos, SeekOrigin.Current);
+            }
+            else
+            {  // The offset of the updated seek pointer is not a legal offset. Loose the buffer.
+                _readPos = _readLen = 0;
+            }
+
+            Contract.Assert(newPos == Position, "newPos (=" + newPos + ") == Position (=" + Position + ")");
+            return newPos;
+        }
+
+        public override void SetLength(Int64 value)
+        {
+            if (value < 0)
+                throw new ArgumentOutOfRangeException("value", SR.ArgumentOutOfRange_NegFileSize);
+            Contract.EndContractBlock();
+
+            EnsureNotClosed();
+            EnsureCanSeek();
+            EnsureCanWrite();
+
+            Flush();
+            _stream.SetLength(value);
+        }
+    }  // class BufferedStream
+}  // namespace
```

Comment:
They absolutely should be in here. I removed them when I removed the Begin/End methods. I took this implementation from NDP\FxCore\src\System.Runtime.WindowsRuntime\System\IO in ProjectN which is the same as the coreclr one except for having USE_ASYNC around the async stuff and also using a different bracket style and other minor things. I removed way too much when I removed the USE_ASYNC tags.

I'll rebase this with one commit including the async stuff and leave out the tests.